ASoC: rt711-sdca: add GE selected mode control

shumingfan opened the pull request:
ASoC: rt711-sdca: add GE selected mode control
This patch adds the 'selected mode' control which overrides the detected_mode.

plbossart commented on sound/soc/codecs/rt711-sdca.c:
```diff
@@ -258,6 +258,8 @@ static int rt711_sdca_headset_detect(struct rt711_sdca_priv *rt711)
 	case 0x00:
 		rt711->jack_type = 0;
+		regmap_write(rt711->regmap,
+			SDW_SDCA_CTL(FUNC_NUM_JACK_CODEC, RT711_SDCA_ENT_GE49, RT711_SDCA_CTL_SELECTED_MODE, 0), 0);
```
i am a bit confused by this patch @shumingfan 

Shouldn't the code override det_mode and change the value so that the updated selected_mode is written line 276?

plbossart replied:
I am also confused by the rest of the switch case where the jack_type would be set based on the det_mode, not selected_mode.

plbossart commented on sound/soc/codecs/rt711-sdca.c:
```diff
@@ -790,6 +792,59 @@ static int rt711_sdca_fu0f_capture_put(struct snd_kcontrol *kcontrol,
+		break;
+	}
+
+	return snd_soc_put_enum_double(kcontrol, ucontrol);
```
still not following. Even if we used a changed in selected_mode as the trigger to update the configuration, the jack_type wouldn't change.

My maybe naive proposal was that the jack detection would be set with the alsa control when the card is created, and whenever an actual jack is detected we would use the value of the alsa control to force the configuration and ignore the detected mode.

shumingfan replied:
@plbossart Could we use the module parameter to force the jack type?
I updated the patch. Please take a look.

plbossart reviewed (commented):
Thanks @shumingfan, I am afraid I don't fully understand the logic behind this patch, see comments below.

plbossart commented on sound/soc/codecs/rt711-sdca.c:
```diff
@@ -267,6 +289,17 @@ static int rt711_sdca_headset_detect(struct rt711_sdca_priv *rt711)
+			det_mode = ge_mode_override;
+			rt711->jack_type =
+				(rt711->jack_type == SND_JACK_HEADPHONE) ? SND_JACK_HEADSET : SND_JACK_HEADPHONE;
+			rt711_sdca_ge_force_jack_type(rt711, det_mode);
```
optimization: I think you could move this line 297 after the if, if det_mode is zero then using rt711_sdca_ge_force_jack_type(rt711, det_mode); would work just fine.

plbossart commented on sound/soc/codecs/rt711-sdca.c:
```diff
@@ -267,6 +289,17 @@ static int rt711_sdca_headset_detect(struct rt711_sdca_priv *rt711)
+			rt711_sdca_ge_force_jack_type(rt711, det_mode);
+		}
+		if (!det_mode)
+			rt711_sdca_ge_force_jack_type(rt711, 0);
```
these two lines can be removed and replaced by rt711_sdca_ge_force_jack_type(rt711, det_mode);

plbossart reviewed (commented):
Thanks @shumingfan. I had the idea of using an alsa control but the module parameter is fine to start testing.

@fredoh9 @marc-hb can you try to see if this patch improve the alsa-bat results with the option 
options snd_soc_rt711_sdca detected_mode=5

fredoh9 commented:
I was able to test with this PR and the module param.

`options snd_soc_rt711_sdca detected_mode=5`

Looks detected_mode is consistent throughout multiple test, however test results have more failure than pass. Captured wave file has very low volume, not even tell whether it's sinewave. Looking still looking at it. i.e increasing gain. Will update.


```
$ grep detected_mode 20240506_dmesg01_pr4969_detected_mode_add_alsabat_tests.txt
[    6.041556] snd_soc_rt711_sdca:rt711_sdca_headset_detect: rt711-sdca sdw:0:0:025d:0711:01: rt711_sdca_headset_detect, detected_mode=0x0
[    6.440514] snd_soc_rt711_sdca:rt711_sdca_headset_detect: rt711-sdca sdw:0:0:025d:0711:01: rt711_sdca_headset_detect, detected_mode=0x5
[   43.739315] snd_soc_rt711_sdca:rt711_sdca_headset_detect: rt711-sdca sdw:0:0:025d:0711:01: rt711_sdca_headset_detect, detected_mode=0x5
[   43.774153] snd_soc_rt711_sdca:rt711_sdca_headset_detect: rt711-sdca sdw:0:0:025d:0711:01: rt711_sdca_headset_detect, detected_mode=0x5
[  177.001368] snd_soc_rt711_sdca:rt711_sdca_headset_detect: rt711-sdca sdw:0:0:025d:0711:01: rt711_sdca_headset_detect, detected_mode=0x5
[  204.427909] snd_soc_rt711_sdca:rt711_sdca_headset_detect: rt711-sdca sdw:0:0:025d:0711:01: rt711_sdca_headset_detect, detected_mode=0x5
[  964.759599] snd_soc_rt711_sdca:rt711_sdca_headset_detect: rt711-sdca sdw:0:0:025d:0711:01: rt711_sdca_headset_detect, detected_mode=0x5
```

marc-hb commented:
> Headset Mic Jack also stays ON.

From the narrow perspective of this PR, this is a PASS! We can discuss and debug alsabat and other issues elsewhere.

Could you test basic, manual hotplug with a sample of headphones and headsets? `evtest` is the most convenient for this.

marc-hb commented:
> I had the idea of using an alsa control but the module parameter is fine to start testing.

A module parameter will be less flexible and less convenient for testing.

shumingfan commented:
@marc-hb @plbossart  I will update the patch, creating an alsa control to change the jack type.

plbossart commented on sound/soc/codecs/rt711-sdca.c:
```diff
@@ -790,6 +817,51 @@ static int rt711_sdca_fu0f_capture_put(struct snd_kcontrol *kcontrol,
+		return -EINVAL;
+	val = snd_soc_enum_item_to_val(e, item[0]) << e->shift_l;
+	rt711->ge_mode_override = val;
+	return 0;
```
isn't there something about the return value of the 'put' that informs upper layers that the value was indeed changed?

shumingfan replied:
OK, will fix

plbossart commented on sound/soc/codecs/rt711-sdca.c:
```diff
@@ -267,6 +285,15 @@ static int rt711_sdca_headset_detect(struct rt711_sdca_priv *rt711)
 	}
 
+	if (rt711->ge_mode_override) {
+		if ((det_mode != rt711->ge_mode_override) && det_mode) {
```
what happens if 'det_mode == 0'? Does this mean nothing is plugged?

shumingfan replied:
Yes

plbossart reviewed (commented):
Looks good to me @shumingfan, two questions below:

marc-hb commented on sound/soc/codecs/rt711-sdca.c:
```diff
@@ -824,6 +896,8 @@ static const struct snd_kcontrol_new rt711_sdca_snd_controls[] = {
 		SDW_SDCA_CTL(FUNC_NUM_MIC_ARRAY, RT711_SDCA_ENT_PLATFORM_FU15, RT711_SDCA_CTL_FU_CH_GAIN, CH_R),
 		8, 3, 0,
 		rt711_sdca_set_gain_get, rt711_sdca_set_gain_put, mic_vol_tlv),
+	SOC_ENUM_EXT("GE49 Selected Mode", rt711_sdca_ge_mode_enum,
```
Is "GE49 Selected Mode" how this will appear in the user interface? "Selected" and "Mode" are very vague and I'm afraid most users will have no idea what "GE49" means.

Can we have one or two more specific words like "Jack type" or "detection" or "microphone" or "headset"?

plbossart replied:
In the SDCA context "GE Selected Mode" makes perfect sense.
It's been a while that users can't figure out what the controls mean...

shumingfan committed:
ASoC: rt711-sdca: add GE selected mode control

This patch adds the 'selected mode' control which overrides the detected_mode.

Signed-off-by: Shuming Fan <shumingf@realtek.com>

plbossart reviewed (approved):
LGTM @shumingfan 

Can you send to Mark Brown, maybe with an updated commit message such as:

The SDCA spec defines a 'selected_mode' control which can override the 'detected_mode' reported by hardware. This is useful for platform integration as well as in cases where the hardware is not able to accurately detect the jack type.

fredoh9 reviewed (approved):
I was able test test with LNL RVP, the overriding mode worked fine.

shumingfan commented:
> LGTM @shumingfan
> 
> Can you send to Mark Brown, maybe with an updated commit message such as:
> 
> The SDCA spec defines a 'selected_mode' control which can override the 'detected_mode' reported by hardware. This is useful for platform integration as well as in cases where the hardware is not able to accurately detect the jack type.

plbossart commented:
shared upstream, thanks @shumingfan

marc-hb commented:
@plbossart don't we need this in this branch already to unblock investigation of other capture issues?

plbossart commented:
good point @marc-hb. I was planning to pick it up at the next merge, but that merge might be a ways off since we have nothing to send. I'll merge it to topic/sof-dev.

marc-hb commented:
Not sure this override is enough:
- https://github.com/thesofproject/sof/issues/9165
